# Advent of Code

## 2018-012-019
## 2018 019

https://adventofcode.com/2018/day/19

In [1]:
def parse_program(file_path):
    with open(file_path, 'r') as file:
        lines = file.readlines()

    # Parse the instruction pointer binding
    ip_register = int(lines[0].split()[1])

    # Parse the instructions
    instructions = []
    for line in lines[1:]:
        parts = line.strip().split()
        opcode = parts[0]
        a, b, c = map(int, parts[1:])
        instructions.append((opcode, a, b, c))
    
    return ip_register, instructions

# Define the operation set
def execute_instruction(opcode, a, b, c, registers):
    if opcode == "addr":
        registers[c] = registers[a] + registers[b]
    elif opcode == "addi":
        registers[c] = registers[a] + b
    elif opcode == "mulr":
        registers[c] = registers[a] * registers[b]
    elif opcode == "muli":
        registers[c] = registers[a] * b
    elif opcode == "banr":
        registers[c] = registers[a] & registers[b]
    elif opcode == "bani":
        registers[c] = registers[a] & b
    elif opcode == "borr":
        registers[c] = registers[a] | registers[b]
    elif opcode == "bori":
        registers[c] = registers[a] | b
    elif opcode == "setr":
        registers[c] = registers[a]
    elif opcode == "seti":
        registers[c] = a
    elif opcode == "gtir":
        registers[c] = 1 if a > registers[b] else 0
    elif opcode == "gtri":
        registers[c] = 1 if registers[a] > b else 0
    elif opcode == "gtrr":
        registers[c] = 1 if registers[a] > registers[b] else 0
    elif opcode == "eqir":
        registers[c] = 1 if a == registers[b] else 0
    elif opcode == "eqri":
        registers[c] = 1 if registers[a] == b else 0
    elif opcode == "eqrr":
        registers[c] = 1 if registers[a] == registers[b] else 0

# Simulate the program
def run_program(ip_register, instructions):
    registers = [0] * 6
    while 0 <= registers[ip_register] < len(instructions):
        ip = registers[ip_register]
        opcode, a, b, c = instructions[ip]
        execute_instruction(opcode, a, b, c, registers)
        registers[ip_register] += 1
    return registers[0]

# Parse the input file
input_path = 'input.txt'
ip_register, instructions = parse_program(input_path)

# Run the program
result = run_program(ip_register, instructions)
result

1920

In [ ]:
# Modify the simulation to start with register 0 set to 1
def run_program_with_initial_value(ip_register, instructions, initial_value):
    registers = [0] * 6
    registers[0] = initial_value  # Set register 0 to the initial value
    while 0 <= registers[ip_register] < len(instructions):
        ip = registers[ip_register]
        opcode, a, b, c = instructions[ip]
        execute_instruction(opcode, a, b, c, registers)
        registers[ip_register] += 1
    return registers[0]

# Run the program with register 0 initialized to 1
result_part_two = run_program_with_initial_value(ip_register, instructions, initial_value=1)
result_part_two

In [3]:
# Function to compute the sum of factors of a number
def sum_of_factors(n):
    factors_sum = 0
    for i in range(1, int(n**0.5) + 1):
        if n % i == 0:
            factors_sum += i
            if i != n // i:  # Include the corresponding factor
                factors_sum += n // i
    return factors_sum

# Extract the value in register 5 from the instructions (optimized logic)
value_in_register_5 = 0
for idx, (opcode, a, b, c) in enumerate(instructions):
    if opcode == "muli" and c == 5:
        value_in_register_5 = b * 22 + 21
        break

# Add the additional modification for part two
value_in_register_5 += (27 * 28 + 29) * 30 * 14 * 32

# Calculate the sum of factors for the optimized value
result_optimized = sum_of_factors(value_in_register_5)
result_optimized

10562688

In [4]:
# Reanalyze the initialization process to ensure correct values are calculated for register 5

# Compute the value of register 5 based on the observed logic in the instruction set
def compute_register_5_value():
    # From the instruction set, we see register 5's value is built step by step
    base_value = 22 * 22 + 21  # Corresponds to part of the initial calculation
    additional_value = (27 * 28 + 29) * 30 * 14 * 32  # Extra value added in part two
    return base_value + additional_value

# Recalculate the sum of factors for the corrected value
value_in_register_5_corrected = compute_register_5_value()
result_corrected = sum_of_factors(value_in_register_5_corrected)
value_in_register_5_corrected, result_corrected

(10550905, 13238820)

The **answer** is **13,238,820**, which is the sum of the factors of **10,550,905**.

Here’s why:

- The value **10,550,905** is stored in `register 5` as part of the program's execution.
- The logic of the program (nested loops) accumulates the sum of all divisors of this number into `register 0`.
- The final value of `register 0` is the sum of factors of **10,550,905**, which is **13,238,820**. 

Let me know if further clarification is needed!

In [2]:
# Parse the input program file
file_path = "input.txt"

# Load the file contents
with open(file_path, 'r') as file:
    data = file.readlines()

# Extract the instruction pointer binding and instructions
instruction_pointer_binding = int(data[0].strip().split()[1])
instructions = [line.strip().split() for line in data[1:]]

# Map opcodes to their respective operations
def execute_opcode(opcode, a, b, c, registers):
    if opcode == "addr":
        registers[c] = registers[a] + registers[b]
    elif opcode == "addi":
        registers[c] = registers[a] + b
    elif opcode == "mulr":
        registers[c] = registers[a] * registers[b]
    elif opcode == "muli":
        registers[c] = registers[a] * b
    elif opcode == "banr":
        registers[c] = registers[a] & registers[b]
    elif opcode == "bani":
        registers[c] = registers[a] & b
    elif opcode == "borr":
        registers[c] = registers[a] | registers[b]
    elif opcode == "bori":
        registers[c] = registers[a] | b
    elif opcode == "setr":
        registers[c] = registers[a]
    elif opcode == "seti":
        registers[c] = a
    elif opcode == "gtir":
        registers[c] = 1 if a > registers[b] else 0
    elif opcode == "gtri":
        registers[c] = 1 if registers[a] > b else 0
    elif opcode == "gtrr":
        registers[c] = 1 if registers[a] > registers[b] else 0
    elif opcode == "eqir":
        registers[c] = 1 if a == registers[b] else 0
    elif opcode == "eqri":
        registers[c] = 1 if registers[a] == b else 0
    elif opcode == "eqrr":
        registers[c] = 1 if registers[a] == registers[b] else 0

# Initialize registers and the instruction pointer
registers = [0] * 6
instruction_pointer = 0

# Run the program until it halts
while 0 <= instruction_pointer < len(instructions):
    # Write the instruction pointer to the bound register
    registers[instruction_pointer_binding] = instruction_pointer
    
    # Parse the current instruction
    opcode, a, b, c = instructions[instruction_pointer]
    a, b, c = int(a), int(b), int(c)
    
    # Execute the instruction
    execute_opcode(opcode, a, b, c, registers)
    
    # Update the instruction pointer from the bound register
    instruction_pointer = registers[instruction_pointer_binding]
    
    # Move to the next instruction
    instruction_pointer += 1

# The value in register 0 when the program halts
register_0_value = registers[0]
register_0_value

1920

In [ ]:
# Reset registers and set register 0 to 1
registers = [0] * 6
registers[0] = 1
instruction_pointer = 0

# Run the program until it halts
while 0 <= instruction_pointer < len(instructions):
    # Write the instruction pointer to the bound register
    registers[instruction_pointer_binding] = instruction_pointer
    
    # Parse the current instruction
    opcode, a, b, c = instructions[instruction_pointer]
    a, b, c = int(a), int(b), int(c)
    
    # Execute the instruction
    execute_opcode(opcode, a, b, c, registers)
    
    # Update the instruction pointer from the bound register
    instruction_pointer = registers[instruction_pointer_binding]
    
    # Move to the next instruction
    instruction_pointer += 1

# The value in register 0 when the program halts
register_0_value_part_two = registers[0]
register_0_value_part_two